### PR TITLE
[GEP-20] Configure high-availability settings for `gardenlet`

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area high-availability
/kind enhancement

**What this PR does / why we need it**:
This PR adds HA options for the `gardenlet` Helm chart as well as to the deployment procedure through `managedseeds`.

**Which issue(s) this PR fixes**:
Fixes parts of #6529

**Release note**:

```feature operator
Gardenlet can now be deployed with multiple replicas and a `failureToleranceType` of either `node` or `zone`. This is supported by the `gardenlet` Helm chart as well as through deployment options in `managedseed` objects. The replica spread is implemented via [TopologySpreadConstraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/).
```

### CODE REVIEW

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl [hunk 1]
```diff
@@ -36,6 +36,35 @@ gardenlet-imagevector-overwrite-components-{{ include "gardenlet.imagevector-ove
...
+{{- end }}
+{{- else }}
+- maxSkew: 1
+  topologyKey: kubernetes.io/hostname
+  whenUnsatisfiable: ScheduleAnyway
+  labelSelector:
+    matchLabels:
+{{ include "gardenlet.deployment.labels" . | indent 8 }}
+{{- end }}
+{{- end }}
```
Comment (plkokanov) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl:65]: I think you are missing one `{{- end }}` here corresponding to `{{- if gt (int .Values.global.gardenlet.replicaCount) 1 -}}`

File: docs/deployment/deploy_gardenlet_manually.md [hunk 2]
```diff
@@ -345,6 +345,18 @@ For all supported infrastructure providers, see [Known Extension Implementations
           type: <provider>
     ```
 
+### Optional: Enable HA mode
+
+You may consider running Gardenlet with multiple replicas, especially if the seed cluster is supposed to host [HA shoot contorl planes](../usage/shoot_high_availability.md).
```
Comment (plkokanov) [inline docs/deployment/deploy_gardenlet_manually.md]: Nit:
```suggestion
You may consider running `gardenlet` with multiple replicas, especially if the seed cluster is configured to host [HA shoot contorl planes](../usage/shoot_high_availability.md).
```

File: docs/deployment/deploy_gardenlet_manually.md [hunk 3]
```diff
@@ -345,6 +345,18 @@ For all supported infrastructure providers, see [Known Extension Implementations
...
+Therefore, the following Helm chart values define the degree of high availability you want to bring to your Gardenlet deployment.
```
Comment (plkokanov) [inline docs/deployment/deploy_gardenlet_manually.md]:
```suggestion
Therefore, the following Helm chart values define the degree of high availability you want to bring to your `gardenlet` deployment.

File: docs/usage/managed_seed.md [hunk 4]
```diff
@@ -21,6 +21,7 @@ Note the following important aspects:
 * The `ManagedSeed` resource must always refer to an existing shoot. Attempting to create a `ManagedSeed` referring to a non-existing shoot will fail.
 * A shoot that is being referred to by a `ManagedSeed` cannot be deleted. Attempting to delete such a shoot will fail.
 * You can omit practically everything from the `seedTemplate` or `gardenlet` section, including all or most of the `Seed` spec fields. Proper defaults will be supplied in all cases, based either on the most common use cases or the information already available in the `Shoot` resource.
+* Also, if your seed is supposed to host HA shoot control planes, then Gardenlet will by default be deployed with multiple replicas across nodes or availability zones.
```
Comment (plkokanov) [inline docs/usage/managed_seed.md]:
```suggestion
* Also, if your seed is configured to host HA shoot control planes, then `gardenlet` will by default be deployed with multiple replicas across nodes or availability zones.
```

File: pkg/apis/seedmanagement/types_managedseed.go [hunk 5]
```diff
@@ -117,6 +118,9 @@ type GardenletDeployment struct {
 	Env []corev1.EnvVar
 	// VPA specifies whether to enable VPA for gardenlet. Defaults to true.
 	VPA *bool
+	// FailureToleranceType determines how gardenlet replicas are spread across the failure domains, either `node` or `zone`.
```
Comment (plkokanov) [inline pkg/apis/seedmanagement/types_managedseed.go]: Nit:
```suggestion
	// FailureToleranceType determines how gardenlet replicas are spread across the failure domains, possible values are either `node` or `zone`.
```

File: charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml [hunk 6]
```diff
@@ -13,6 +13,13 @@ reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-garde
 reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-seed.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.kubeconfig-seed.name" . }}
 {{- end }}
 {{- end -}}
+{{- define "gardenlet.deployment.labels" -}}
+app: gardener
+role: gardenlet
+chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
```
Comment (plkokanov) [inline charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml:24]: With this the "helm-specific" labels will also be added to the `topologySpreadConstraints`, previously when we used `podAntiAffinity`, only the `app` and `role` labels were used. Are the helm labels necessary?

File: pkg/gardenlet/controller/managedseed/charttest/charttest.go [hunk 7]
```diff
@@ -1015,64 +1015,96 @@ func ComputeExpectedGardenletDeploymentSpec(
...
+				},
+				WhenUnsatisfiable: corev1.ScheduleAnyway,
+			})
 		}
+	}
 
-		if deploymentConfiguration.PodAnnotations != nil {
-			deployment.Template.ObjectMeta.Annotations = utils.MergeStringMaps(deployment.Template.ObjectMeta.Annotations, deploymentConfiguration.PodAnnotations)
-		}
+	if deploymentConfiguration.Env != nil {
```
Comment (plkokanov) [inline pkg/gardenlet/controller/managedseed/charttest/charttest.go]: This and the other checks below were moved so that they are now outside the block that checks 
`if deploymentConfiguration != nil {` which could cause nil pointer exceptions.

File: pkg/gardenlet/controller/managedseed/charttest/charttest.go [hunk 8]
```diff
@@ -1015,64 +1015,96 @@ func ComputeExpectedGardenletDeploymentSpec(
...
+
+		topologySpreadConstraintLabels := map[string]string{
+			"app":      "gardener",
+			"role":     "gardenlet",
+			"chart":    "runtime-0.1.0",
+			"release":  "gardenlet",
+			"heritage": "Tiller",
+		}
+
+		if failureToleranceType := deploymentConfiguration.FailureToleranceType; failureToleranceType != nil {
```
Comment (plkokanov) [inline pkg/gardenlet/controller/managedseed/charttest/charttest.go]: Shouldn't this be executed only if the `deployment.Replicas > 1`? From the `_helpers.tpl` I see that the `gardenlet.deployment.topologySpreadConstraints` only has content if `.Values.global.gardenlet.replicaCount` is greater than 1. Adding a test entry that looks like this under https://github.com/gardener/gardener/blob/9a0b8d5375f49dd120db559df928e0236f2ae366/pkg/gardenlet/controller/managedseed/charttest/gardenlet_chart_test.go#L399-L402:
```
Entry("verify deployment with replica count 1 and `zone` failureTolerance", nil, nil, nil, nil, nil, nil, &seedmanagement.GardenletDeployment{
			ReplicaCount:         pointer.Int32(1),
			FailureToleranceType: &failureToleranceZone,
		}, nil, nil, nil, map[string]string{"gardenlet-configmap": "gardenlet-configmap-a99e2d4b"}),
```
returns an error.

File: pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go [hunk 9]
```diff
@@ -343,13 +352,43 @@ var _ = Describe("ValuesHelper", func() {
...
 		})
 
 		It("should compute the correct gardenlet chart values without bootstrap", func() {
 			result, err := vh.GetGardenletChartValues(mergedDeployment, mergedGardenletConfig(false), "")
 			Expect(err).ToNot(HaveOccurred())
-			Expect(result).To(Equal(gardenletChartValues(false, "")))
+			Expect(result).To(Equal(gardenletChartValues(false, "", "node", 1, nil)))
+		})
+
+		It("should compute the correct gardenlet chart values if seed config is multi-zonal", func() {
```
Comment (plkokanov) [inline pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go]: Should we add some more tests that check the following configurations: ReplicaCount > 1, FailureToleranceType = "node" (checks if "node" is preserved since it is already set); ReplicaCount=1, FailureToleranceType = nil (checks that `failureToleranceType` value is not added)?

**Review by plkokanov — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl (hunk 1 above)
Comment (timuthy) [reply charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl:65]: Hm, are you sure? The corresponding `{{- end }}` is already in line `65`?

Comment (plkokanov) [reply charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl:65]: Ah sorry, forgot to remove this comment before posting the review.

**Comment by timuthy** [comment]
Thanks for your rewiew and valuable feedback @plkokanov 🙂 I addressed all of you comments, PTAL.

File: docs/usage/managed_seed.md (hunk 4 above)
Comment (plkokanov) [reply docs/usage/managed_seed.md]: Hmm, this nit wasn't addressed

File: pkg/apis/seedmanagement/v1alpha1/types_managedseed.go [hunk 10]
```diff
@@ -141,6 +141,10 @@ type GardenletDeployment struct {
 	// VPA specifies whether to enable VPA for gardenlet. Defaults to true.
 	// +optional
 	VPA *bool `json:"vpa,omitempty" protobuf:"bytes,11,rep,name=vpa"`
+	// FailureToleranceType determines how gardenlet replicas are spread across the failure domains, either `node` or `zone`.
```
Comment (plkokanov) [inline pkg/apis/seedmanagement/v1alpha1/types_managedseed.go]: Could you align this to: https://github.com/gardener/gardener/blob/9f37f495ed44a5642333abefae0cfe87a79e16e6/pkg/apis/seedmanagement/types_managedseed.go#L121

File: pkg/apis/seedmanagement/types_managedseed.go [hunk 11]
```diff
@@ -117,6 +118,9 @@ type GardenletDeployment struct {
 	Env []corev1.EnvVar
 	// VPA specifies whether to enable VPA for gardenlet. Defaults to true.
 	VPA *bool
+	// FailureToleranceType determines how gardenlet replicas are spread across the failure domains, possible values are either `node` or `zone`.
+	// Please make sure to adjust the replicaCount accordingly if you intend to run an HA setup for gardenlet.
+	FailureToleranceType *gardencorev1beta1.FailureToleranceType
```
Comment (plkokanov) [inline pkg/apis/seedmanagement/types_managedseed.go]: Should validation be added for this field to make sure that only `node` and `zone` can be specified? When deploying `gardenlet` as part of a `ManagedSeed` that has `failureToleranceType: asdf` and more than one replicas the following is added to the `deployment`:
```
topologySpreadConstraints:
- labelSelector:
    matchLabels:
      app: gardener
      role: gardenlet
  maxSkew: 1
  topologyKey: kubernetes.io/hostname
  whenUnsatisfiable: ScheduleAnyway
```
This is technically correct, but might be a bit unexpected.

As an additional remark, the only way I can add this entry for the `topologySpreadConstraints` is if I specify a random `failureToleranceType` and replicas > 1. 
Do you think it makes sense to add a third `failureToleranceType`, e.g. `failureToleranceType: none` with which to make it possible/clearer to add such an entry to the `topologySpreadConstraints` of the `gardenlet` deployment (note that I might be missing some context here from the HA topic).

File: example/55-managedseed-gardenlet.yaml [hunk 12]
```diff
@@ -12,7 +12,8 @@ spec:
     bootstrap: ServiceAccount # Mechanism that should be used for bootstrapping gardenlet connection to the Garden cluster, one of ServiceAccount, BootstrapToken, None
     mergeWithParent: true # If true, the deployment parameters and GardenletConfiguration of the parent gardenlet will be merged with the specified deployment parameters and GardenletConfiguration
 #   deployment: # gardenlet deployment parameters
-#     replicaCount: 1      
+#     replicaCount: 1
+#     failureTolerance: zone # one of `node` or `zone`
```
Comment (plkokanov) [inline example/55-managedseed-gardenlet.yaml]:
```suggestion
#     failureToleranceType: zone # one of `node` or `zone`
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Tested the PR by creating an HA `ManagedSeed` and setting various values for the `gardenlet` replicas and the `failureToleranceType`. It seems to work fine.
I just have a couple more comments.

File: pkg/apis/seedmanagement/types_managedseed.go (hunk 11 above)
Comment (timuthy) [reply pkg/apis/seedmanagement/types_managedseed.go]: > As an additional remark, the only way I can add this entry for the topologySpreadConstraints is if I specify a random failureToleranceType and replicas > 1.

You should especially see this entry if no `failureToleranceType` is specified which is the equivalent to the "soft" anti-affinity configuration that was used earlier. I'd rather refrain from explicitly having `failureToleranceType: none` as we don't support this type neither for the shoot control plane (via shoot spec), nor for the seed.

Comment (timuthy) [reply pkg/apis/seedmanagement/types_managedseed.go]: Validation makes sense, indeed 👍

Comment (plkokanov) [reply pkg/apis/seedmanagement/types_managedseed.go]: > You should especially see this entry if no `failureToleranceType` is specified which is the equivalent to the "soft" anti-affinity configuration that was used earlier. I'd rather refrain from explicitly having `failureToleranceType: none` as we don't support this type neither for the shoot control plane (via shoot spec), nor for the seed.

I see, the `failureToleranceType` is only defaulted to `zone` if the seed is specified to be HA in the `seedConfig`. I guess `failureToleranceType!=nil` is something that should be enforced for HA seeds (as it is in the current PR). Then indeed it doesn't make sense to add a `none` `failureToleranceType`.

**Comment by timuthy** [comment]
Thanks @plkokanov. I addressed your feedback and also added proper defaulting for the gardenlet deployment in [b1bad45](https://github.com/gardener/gardener/pull/6750/commits/b1bad45f041bcf46ba6f0d09d8bf0b5085eca33b) which I missed earlier.

File: charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml [hunk 13]
```diff
@@ -13,6 +13,16 @@ reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-garde
 reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-seed.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.kubeconfig-seed.name" . }}
 {{- end }}
 {{- end -}}
+{{- define "gardenlet.deployment.matchLabels" -}}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml:17]: Nit: We could leave 1 empty line between the different `define`s to make them more readable:
```yaml
{{- define "gardenlet.deployment.annotations" -}}
reference.resources.gardener.cloud/configmap-{{ include "gardenlet.config.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.config.name" . }}
{{- if .Values.global.gardenlet.imageVectorOverwrite }}
reference.resources.gardener.cloud/configmap-{{ include "gardenlet.imagevector-overwrite.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.imagevector-overwrite.name" . }}
{{- end }}
{{- if .Values.global.gardenlet.componentImageVectorOverwrites }}
reference.resources.gardener.cloud/configmap-{{ include "gardenlet.imagevector-overwrite-components.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.imagevector-overwrite-components.name" . }}
{{- end }}
{{- if .Values.global.gardenlet.config.gardenClientConnection.kubeconfig }}
reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-garden.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.kubeconfig-garden.name" . }}
{{- end }}
{{- if .Values.global.gardenlet.config.seedClientConnection.kubeconfig }}
reference.resources.gardener.cloud/secret-{{ include "gardenlet.kubeconfig-seed.name" . | sha256sum | trunc 8 }}: {{ include "gardenlet.kubeconfig-seed.name" . }}
{{- end }}
{{- end -}}

{{- define "gardenlet.deployment.matchLabels" -}}
app: gardener
role: gardenlet
{{- end -}}

{{- define "gardenlet.deployment.labels" -}}
{{- include "gardenlet.deployment.matchLabels" . }}
chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
release: "{{ .Release.Name }}"
heritage: "{{ .Release.Service }}"
{{- end -}}
```

File: charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml [hunk 14]
```diff
@@ -50,38 +55,19 @@ spec:
...
-        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
-        release: "{{ .Release.Name }}"
-        heritage: "{{ .Release.Service }}"
+{{ include "gardenlet.deployment.labels" . | indent 8 }}
         projected-token-mount.resources.gardener.cloud/skip: "true"
         {{- if .Values.global.gardenlet.podLabels }}
 {{ toYaml .Values.global.gardenlet.podLabels | indent 8 }}
         {{- end }}
     spec:
+{{- include "gardenlet.deployment.topologySpreadConstraints" . | nindent 6 }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/deployment.yaml]: For consistency reasons with the rest of the template file:
```suggestion
{{ include "gardenlet.deployment.topologySpreadConstraints" . | indent 6 }}
```

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl [hunk 15]
```diff
@@ -36,6 +36,35 @@ gardenlet-imagevector-overwrite-components-{{ include "gardenlet.imagevector-ove
...
+    matchLabels:
+{{ include "gardenlet.deployment.matchLabels" . | indent 8 }}
+{{- end }}
+{{- else }}
+- maxSkew: 1
+  topologyKey: kubernetes.io/hostname
+  whenUnsatisfiable: ScheduleAnyway
+  labelSelector:
+    matchLabels:
+{{ include "gardenlet.deployment.matchLabels" . | indent 8 }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl]:
```suggestion
{{ include "gardenlet.deployment.matchLabels" . | indent 6 }}
```

When I checked with `helm template` the labels were rendered as:
```yaml
        labelSelector:
          matchLabels:
              app: gardener
              role: gardenlet
```

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl [hunk 16]
```diff
@@ -36,6 +36,35 @@ gardenlet-imagevector-overwrite-components-{{ include "gardenlet.imagevector-ove
...
+{{- if eq .Values.global.gardenlet.failureToleranceType "zone" }}
+- maxSkew: 1
+  topologyKey: topology.kubernetes.io/zone
+  whenUnsatisfiable: DoNotSchedule
+  labelSelector:
+    matchLabels:
+{{ include "gardenlet.deployment.matchLabels" . | indent 8 }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl]:
```suggestion
{{ include "gardenlet.deployment.matchLabels" . | indent 6 }}
```

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl [hunk 17]
```diff
@@ -36,6 +36,35 @@ gardenlet-imagevector-overwrite-components-{{ include "gardenlet.imagevector-ove
...
+{{- define "gardenlet.deployment.topologySpreadConstraints" -}}
+{{- if gt (int .Values.global.gardenlet.replicaCount) 1 -}}
+topologySpreadConstraints:
+{{- if or (eq .Values.global.gardenlet.failureToleranceType "node") (eq .Values.global.gardenlet.failureToleranceType "zone") }}
+- maxSkew: 1
+  topologyKey: kubernetes.io/hostname
+  whenUnsatisfiable: DoNotSchedule
+  labelSelector:
+    matchLabels:
+{{ include "gardenlet.deployment.matchLabels" . | indent 8 }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl]:
```suggestion
{{ include "gardenlet.deployment.matchLabels" . | indent 6 }}
```

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl (hunk 15 above)
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl]: When is this `else` statement executed?

File: pkg/apis/core/validation/utils.go [hunk 18]
```diff
@@ -110,3 +113,20 @@ func getPercentValue(intOrStringValue intstr.IntOrString) (int, bool) {
 	value, _ := strconv.Atoi(intOrStringValue.StrVal[:len(intOrStringValue.StrVal)-1])
 	return value, true
 }
+
+var availableFailureTolerance = sets.NewString(
+	string(core.FailureToleranceTypeNode),
+	string(core.FailureToleranceTypeZone),
+)
+
+// ValidateFailureToleranceValue validates if the FailureTolerance has a supported value
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/utils.go]:
```suggestion
// ValidateFailureToleranceTypeValue validates if the FailureTolerance has a supported value
```

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl (hunk 15 above)
Comment (rfranzke) [reply charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl]: I think it's executed when `failureToleranceType` is neither `node` nor `zone` IIUC 👀

File: docs/deployment/deploy_gardenlet_manually.md [hunk 19]
```diff
@@ -345,6 +345,18 @@ For all supported infrastructure providers, see [Known Extension Implementations
           type: <provider>
     ```
 
+### Optional: Enable HA mode
+
+You may consider running `gardenlet` with multiple replicas, especially if the seed cluster is configured to host [HA shoot contorl planes](../usage/shoot_high_availability.md).
```
Comment (rfranzke) [inline docs/deployment/deploy_gardenlet_manually.md]:
```suggestion
You may consider running `gardenlet` with multiple replicas, especially if the seed cluster is configured to host [HA shoot control planes](../usage/shoot_high_availability.md).
```

File: docs/deployment/deploy_gardenlet_manually.md [hunk 20]
```diff
@@ -345,6 +345,18 @@ For all supported infrastructure providers, see [Known Extension Implementations
...
+Therefore, the following Helm chart values define the degree of high availability you want to bring to your `gardenlet` deployment.
```
Comment (rfranzke) [inline docs/deployment/deploy_gardenlet_manually.md]:
```suggestion
Therefore, the following Helm chart values define the degree of availability you want to achieve for the `gardenlet` deployment.
```

File: docs/usage/managed_seed.md [hunk 21]
```diff
@@ -21,6 +21,7 @@ Note the following important aspects:
 * The `ManagedSeed` resource must always refer to an existing shoot. Attempting to create a `ManagedSeed` referring to a non-existing shoot will fail.
 * A shoot that is being referred to by a `ManagedSeed` cannot be deleted. Attempting to delete such a shoot will fail.
 * You can omit practically everything from the `seedTemplate` or `gardenlet` section, including all or most of the `Seed` spec fields. Proper defaults will be supplied in all cases, based either on the most common use cases or the information already available in the `Shoot` resource.
+* Also, if your seed is configured to host HA shoot control planes, then `gardenlet` will by default be deployed with multiple replicas across nodes or availability zones.
```
Comment (rfranzke) [inline docs/usage/managed_seed.md]:
```suggestion
* Also, if your seed is configured to host HA shoot control planes then `gardenlet` will be deployed with multiple replicas across nodes or availability zones by default.
```

File: pkg/apis/core/validation/utils_test.go [hunk 22]
```diff
@@ -17,9 +17,48 @@ package validation_test
...
+		It("highAvailability is set to an unsupported value", func() {
+			errorList := validation.ValidateFailureToleranceTypeValue("region", fldPath)
+			Expect(errorList).To(ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeNotSupported),
+					"Field": Equal(fldPath.String()),
+				}))))
+		})
+	})
+
```
Comment (rfranzke) [inline pkg/apis/core/validation/utils_test.go]:
```suggestion
```

File: pkg/apis/core/validation/utils_test.go [hunk 23]
```diff
@@ -17,9 +17,48 @@ package validation_test
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
 func makeDurationPointer(d time.Duration) *metav1.Duration {
 	return &metav1.Duration{Duration: d}
 }
+
+var _ = Describe("utils", func() {
+
```
Comment (rfranzke) [inline pkg/apis/core/validation/utils_test.go]:
```suggestion
```

File: pkg/apis/core/validation/utils_test.go [hunk 24]
```diff
@@ -17,9 +17,48 @@ package validation_test
...
+	Context("#ValidateFailureToleranceTypeValue", func() {
+		var fldPath *field.Path
+
+		BeforeEach(func() {
+			fldPath = field.NewPath("spec", "highAvailability", "failureTolerance", "type")
+		})
+
+		It("highAvailability is set to failureTolerance of node", func() {
+			errorList := validation.ValidateFailureToleranceTypeValue(core.FailureToleranceTypeNode, fldPath)
+			Expect(errorList).To(HaveLen(0))
```
Comment (rfranzke) [inline pkg/apis/core/validation/utils_test.go]: Nit: `BeEmpty()` improves readability (same below)

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 25]
```diff
@@ -103,6 +104,19 @@ func setDefaultsGardenlet(obj *Gardenlet, name, namespace string) {
...
+	if gardenletConfig.SeedConfig != nil {
+		if _, ok := gardenletConfig.SeedConfig.Labels[v1beta1constants.LabelSeedMultiZonal]; ok {
+			if obj.Deployment.FailureToleranceType == nil {
+				failureToleranceZone := gardencorev1beta1.FailureToleranceTypeZone
+				obj.Deployment.FailureToleranceType = &failureToleranceZone
+			}
+			if obj.Deployment.ReplicaCount == nil {
+				obj.Deployment.ReplicaCount = pointer.Int32(2)
+			}
+		}
```
Comment (rfranzke) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go:127]: Should we wait for https://github.com/gardener/gardener/pull/6723? If not, can you add a TODO that this new `spec` field needs to be evaluated here as well (and add it to your umbrella issue as new task))?

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed_test.go [hunk 26]
```diff
@@ -140,24 +141,8 @@ var _ = Describe("Defaults", func() {
...
-						SeedConfig: &configv1alpha1.SeedConfig{
-							SeedTemplate: gardencorev1beta1.SeedTemplate{
-								Spec: gardencorev1beta1.SeedSpec{
-									Backup: &gardencorev1beta1.SeedBackup{},
-								},
-							},
-						},
-					}),
-				},
-			}
```
Comment (rfranzke) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed_test.go]: Why was this changed? (same below)

File: pkg/apis/seedmanagement/v1alpha1/types_managedseed.go [hunk 27]
```diff
@@ -120,7 +120,7 @@ type GardenletDeployment struct {
 	// Image is the gardenlet container image.
 	// +optional
 	Image *Image `json:"image,omitempty" protobuf:"bytes,4,opt,name=image"`
-	// Resources are the compute resources required by the gardenlet container.
+	// Resources are the computed resources required by the gardenlet container.
```
Comment (rfranzke) [inline pkg/apis/seedmanagement/v1alpha1/types_managedseed.go]: I think `compute` was intentional. Revert it?

File: pkg/apis/seedmanagement/types_managedseed.go [hunk 28]
```diff
@@ -103,7 +103,7 @@ type GardenletDeployment struct {
 	ServiceAccountName *string
 	// Image is the gardenlet container image.
 	Image *Image
-	// Resources are the compute resources required by the gardenlet container.
+	// Resources are the computed resources required by the gardenlet container.
```
Comment (rfranzke) [inline pkg/apis/seedmanagement/types_managedseed.go]: I think `compute` was intentional. Revert it?

File: pkg/apis/seedmanagement/validation/managedseed_test.go [hunk 29]
```diff
@@ -346,6 +346,34 @@ var _ = Describe("ManagedSeed Validation Tests", func() {
 				))
 			})
 
+			It("should allow valid failure tolerance type zone", func() {
+				failureToleranceTypeZone := core.FailureToleranceTypeZone
+				managedSeed.Spec.Gardenlet.Deployment.FailureToleranceType = &failureToleranceTypeZone
+
+				Expect(ValidateManagedSeed(managedSeed)).To(HaveLen(0))
```
Comment (rfranzke) [inline pkg/apis/seedmanagement/validation/managedseed_test.go]: same here and below

File: pkg/gardenlet/controller/managedseed/managedseed_valueshelper.go [hunk 30]
```diff
@@ -138,6 +139,18 @@ func (vp *valuesHelper) GetGardenletChartValues(
 ) (map[string]interface{}, error) {
 	var err error
 
+	// Set HA related default values if gardenlet is responsible for a multi-zonal seed.
+	if config.SeedConfig != nil {
+		if _, ok := config.SeedConfig.Labels[v1beta1constants.LabelSeedMultiZonal]; ok {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/managedseed/managedseed_valueshelper.go]: Same question about https://github.com/gardener/gardener/pull/6723 / a TODO statement here.

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go (hunk 25 above)
Comment (timuthy) [reply pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go:127]: Yes, let's wait as it's more or less in a final state to get merged.

**Comment by timuthy** [comment]
I converted the PR into draft since we want to wait for #6723 to be merged first.

**Comment by timuthy** [comment]
The PR has been rebased and feedback has been addressed, as well. PTAL @rfranzke and @ialidzhikov.

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go [hunk 31]
```diff
@@ -343,14 +352,112 @@ var _ = Describe("ValuesHelper", func() {
...
 		})
 
 		It("should compute the correct gardenlet chart values without bootstrap", func() {
 			result, err := vh.GetGardenletChartValues(mergedDeployment, mergedGardenletConfig(false), "")
 			Expect(err).ToNot(HaveOccurred())
-			Expect(result).To(Equal(gardenletChartValues(false, "")))
+			Expect(result).To(Equal(gardenletChartValues(false, "", "node", 1, nil)))
+		})
+
+		Context("when HA is configured for seed ", func() {
```
Comment (ialidzhikov) [inline pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go]:
```suggestion
		Context("when HA is configured for seed", func() {
```

**Review by ialidzhikov — COMMENTED** [review]
/approve

File: pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go (hunk 31 above)
Comment (timuthy) [reply pkg/gardenlet/controller/managedseed/managedseed_valueshelper_test.go]: Changed in 682377197fb9ada9212326ef3ebb666106b7f81a.

**Review by ialidzhikov — COMMENTED** [review]
/lgtm
